packages/office-ui-fabric-react/src/components/Modal/Modal.base.tsx

```diff
@@ -192,4 +200,17 @@ export class ModalBase extends BaseComponent<IModalProps, IDialogState> implemen
       this.props.onDismissed();
     }
   }
+
+  // Handle onLayerDidMount and layerProps.onLayerDidMount
+  // Can be deleted once IModalProps.onLayerDidMount has been removed
+  private _onLayerDidMount = (): void => {
```

Comment:
Not sure about about this... `layerProps.onLayerDidMount` and `props.layerDidMount` should be mutual exclusive, meaning that in case the user provided both of them neglecting the `@deprecated` flag, they are both going to fire.

Why not filter them out where you build the `layerProps` object you spread on `Layer`. Should look like:
```ts
// Have this in the props destructuring above in the render method.
const { layerProps, onLayerDidMount } = this.props;

const mergedLayerProps = {
    ...DefaultLayerProps,
    ...layerProps,
    onLayerDidMount: layerProps && layerProps.onLayerDidMount ? layerProps.onLayerDidMount : onLayerDidMount
};
```
And later just spread it on layer `<Layer {...mergedLayerProps} >`. This way you let the invocation of the callback inside the `Layer` component where it supposed to happen as opposed to passing a function which when invoked triggers the callback.

Also you could probably use the `BaseComponent` helper `_warnDeprecation()` and `_warnMutuallyExclusive()` methods for a cleaner deprecation pattern. Example: [here](https://github.com/OfficeDev/office-ui-fabric-react/blob/master/packages/experiments/src/components/Shimmer/Shimmer.base.tsx#L33-L43)

Reply:
* Changed `mergedLayerProps` creation to your proposed solution
* Added `_warnDeprecation` for `onLayerDidMount`
* `_warnMutuallyExclusive` seems not to work when new property is a "sub-property". Or did I miss something?
* Due to the new deprecation warning, I had to update Dialog to use the new Property
* Have not found a better solution to merge `onLayerDidMount` into `modalProps.layerProps.onLayerDidMount`